src/cli/subcommands/car_cmd.rs

```diff
@@ -0,0 +1,178 @@
+        car_files: Vec<PathBuf>,
+        /// The output `.car` file path
+        #[arg(short, long)]
+        output: PathBuf,
+    },
+}
+
+impl CarCommands {
+    pub async fn run(self) -> anyhow::Result<()> {
+        match self {
+            Self::Concat { car_files, output } => {
+                let readers: Vec<_> = futures::stream::iter(car_files)
+                    .then(tokio::fs::File::open)
+                    .map_ok(tokio::io::BufReader::new)
+                    .map_ok(tokio::io::BufReader::compat)
+                    .map_err(fvm_ipld_car::Error::from)
+                    .and_then(CarReader::new)
+                    .try_collect()
+                    .await?;
+
+                let roots = {
+                    let mut roots = vec![];
+                    let mut seen = CidHashSet::default();
+                    for reader in &readers {
+                        for &root in &reader.header.roots {
+                            if seen.insert(root) {
+                                roots.push(root);
+                            }
+                        }
+                    }
+                    roots
+                };
+
+                let car_writer = CarHeader::from(roots);
+                let mut output_file =
+                    tokio::io::BufWriter::new(tokio::fs::File::create(output).await?).compat();
+
+                car_writer
+                    .write_stream_async(
+                        &mut output_file,
+                        &mut Box::pin(dedup_block_stream(merge_car_readers(readers))),
+                    )
+                    .await?;
+            }
+        }
+        Ok(())
+    }
+}
+
+fn read_car_as_stream<R>(reader: CarReader<R>) -> impl Stream<Item = BlockPair>
+where
+    R: AsyncRead + Send + Unpin,
+{
+    futures::stream::unfold(reader, move |mut reader| async {
+        reader
+            .next_block()
+            .await
+            .expect("Failed to call CarReader::next_block")
+            .map(|b| ((b.cid, b.data), reader))
+    })
+}
+
+fn merge_car_readers<R>(readers: Vec<CarReader<R>>) -> impl Stream<Item = BlockPair>
+where
+    R: AsyncRead + Send + Unpin,
+{
+    futures::stream::iter(readers).flat_map(read_car_as_stream)
+}
+
+fn dedup_block_stream(stream: impl Stream<Item = BlockPair>) -> impl Stream<Item = BlockPair> {
+    let mut seen = CidHashSet::default();
+    stream.filter(move |(cid, _data)| futures::future::ready(seen.insert(*cid)))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ahash::HashSet;
+    use cid::multihash;
+    use cid::multihash::MultihashDigest;
+    use cid::Cid;
+    use fvm_ipld_car::Block;
+    use fvm_ipld_encoding::DAG_CBOR;
+    use pretty_assertions::assert_eq;
+    use quickcheck::Arbitrary;
+    use quickcheck_macros::quickcheck;
+
+    #[derive(Debug, Clone)]
+    struct Blocks(Vec<Block>);
+
+    impl Blocks {
+        async fn into_car_bytes(self) -> anyhow::Result<Vec<u8>> {
```

Comment:
Panics are nice in tests.

```suggestion
        async fn into_car_bytes(self) -> Vec<u8> {
```